File: 0019-remove-nth-node-from-end-of-list/javascript/Solution.js

// Definition for singly-linked list.
function ListNode(val, next) {
  this.val = (val === undefined ? 0 : val)
  this.next = (next === undefined ? null : next)
}

/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function (head, n) {
  const dummy = new ListNode(0);
  dummy.next = head;
  let p = q = dummy;

  while (n > 0 && q.next) {
    q = q.next;
    n--;
  }

  if (n != 0) {
    return dummy.next;
  }

  while (q.next) {
    p = p.next;
    q = q.next;
  }
  p.next = p.next.next;
  return dummy.next;
};

File: 0019-remove-nth-node-from-end-of-list/Solution.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        if not head:
            return head
        dummy = ListNode(-1)
        dummy.next = head
        p = dummy
        q = dummy

        while n > 0 and q.next:
            q = q.next
            n -= 1

        while q.next:
            p = p.next
            q = q.next

        p.next = p.next.next
        return dummy.next